google/stackdriver_monitoring/stackdriver_handlers.py

```diff
@@ -13,24 +13,110 @@
 # limitations under the License.
 
 import cgi
+import json
 import logging
 from stackdriver_client import StackdriverClient
 
 
-def accepts_html(request):
+def accepts_content_type(request, content_type):
+  """Determine if the request wants the given content_type as a response."""
   if not hasattr(request, 'headers'):
     return False
   accept = request.headers.get('accept', None)
   if not accept:
     return None
-  return 'text/html' in str(accept)
+  return content_type in str(accept)
+
+
+class BatchProcessor(object):
+  """Helper class for managing events in batch."""
+  def __init__(self, project, stackdriver, data_list,
+               invocation_factory, get_name):
+    """Constructor.
+
+    Args:
+      data_list: [object] The data to operate on.
+    """
+    self.__project = project
+    self.__stackdriver = stackdriver
+    self.__data_list = data_list
+    self.__num_data = len(self.__data_list)
+    self.__invocation_factory = invocation_factory
+    self.__get_name = get_name
+
+    self.batch_response = [None] * self.__num_data
+    self.num_bad = 0  # number of unsuccessful responses.
+    self.num_ok = 0   # number of successful responses.
+    self.was_ok = [None] * self.__num_data
+
+  def handle_batch_response(self, index_str, good, bad):
+    """Record an individual response.
+
+    Args:
+      index_str: [string] The index in the batch response corresponds to
+         the index in the original query so the mapping is 1:1.
+      good: [string] The response from the request, which succeeded.
+      bad: [string] The response from the request, which failed.
+    """
+    index = int(index_str)
+    if bad:
+      self.was_ok[index] = False
+      self.num_bad += 1
+      self.batch_response[index] = 'ERROR {0}'.format(cgi.escape(str(bad)))
+      logging.error(bad)
+    else:
+      self.was_ok[index] = True
+      self.num_ok += 1
+      if not good:
+        good = ''
+      self.batch_response[index] = 'OK {0}'.format(cgi.escape(good))
+
+  def process(self):
+    """Process all the data by sending one or more batches."""
+    batch = self.__stackdriver.service.new_batch_http_request()
+    max_batch = 100
+    count = 0
+
+    for data in self.__data_list:
+      invocation = self.__invocation_factory(data)
+      batch.add(invocation, callback=self.handle_batch_response,
+                request_id=str(count))
+      count += 1
+      if count % max_batch == 0:
+        logging.info('Executing batch of %d', max_batch)
+        batch.execute()
+        batch = self.__stackdriver.service.new_batch_http_request()
+
+    if count % max_batch:
+      logging.info('Executing final batch of %d', count % max_batch)
```

Comment:
This logging is potentially misleading, since this won't be hit if `len(self.__data_list) % max_batch == 0`. If the user doesn't see "final batch" in their logging after expecting it, this serves as a red herring.